File: lib/format.dart
import 'models.dart';

const _months = ['JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL', 'AUG', 'SEP', 'OCT', 'NOV', 'DEC'];

String formatOdo(num n) {
  final v = n.round();
  final digits = v.toString();
  final buffer = StringBuffer();
  for (var i = 0; i < digits.length; i++) {
    if (i > 0 && (digits.length - i) % 3 == 0) buffer.write(' ');
    buffer.write(digits[i]);
  }
  return buffer.toString();
}

String formatDate(String? iso) {
  if (iso == null || iso.isEmpty) return '';
  final d = DateTime.tryParse(iso);
  if (d == null) return iso;
  return '${d.day} ${_months[d.month - 1]} ${d.year}';
}

String formatDateHeader([DateTime? d]) {
  final date = d ?? DateTime.now();
  return '${date.day} ${_months[date.month - 1]} ${date.year}';
}

bool isChecked(Object? v) => v == true || v == 'true' || v == 'yes' || v == 'done';

List<FieldDef> fieldsForLog(Vehicle? vehicle, LogEntry? log) {
  if (vehicle == null || log == null) return const [];
  final byId = vehicle.categories.where((c) => c.id == log.categoryId);
  final cat = byId.isNotEmpty ? byId.first : vehicle.categories.where((c) => c.name == log.categoryName).firstOrNull;
  return cat?.fields ?? const [];
}

String formatLogValues(Map<String, Object?>? values, List<FieldDef> fields) {
  if (values == null) return '';
  final byLabel = {for (final f in fields) f.label: f};
  return values.entries
      .where((e) => e.value != null && e.value != '' && e.value != false)
      .map((e) {
        if (isChecked(e.value)) return '${e.key} ✓';
        final unit = byLabel[e.key]?.unit;
        return unit != null ? '${e.key}: ${e.value} $unit' : '${e.key}: ${e.value}';
      })
      .join('  ·  ');
}

String formatLogLine(Vehicle? vehicle, LogEntry? log) {
  return formatLogValues(log?.values, fieldsForLog(vehicle, log));
}

File: src/format.js
const MONTHS = ["JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"];

export function formatOdo(n) {
  const v = Math.round(Number(n) || 0);
  return v.toString().replace(/\B(?=(\d{3})+(?!\d))/g, " ");
}

export function formatDate(iso) {
  if (!iso) return "";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return iso;
  return `${d.getDate()} ${MONTHS[d.getMonth()]} ${d.getFullYear()}`;
}

export function formatDateHeader(d = new Date()) {
  return `${d.getDate()} ${MONTHS[d.getMonth()]} ${d.getFullYear()}`;
}

export function isChecked(v) {
  return v === true || v === "true" || v === "yes" || v === "done";
}

export function fieldsForLog(vehicle, log) {
  if (!vehicle || !log) return [];
  const cats = vehicle.categories || [];
  const cat =
    cats.find((c) => c.id === log.categoryId) || cats.find((c) => c.name === log.categoryName);
  return cat?.fields || [];
}

export function formatLogValues(values, fields) {
  if (!values) return "";
  const byLabel = new Map((fields || []).map((f) => [f.label, f]));
  return Object.entries(values)
    .filter(([, v]) => v != null && v !== "" && v !== false)
    .map(([k, v]) => {
      if (isChecked(v)) return `${k} ✓`;
      const unit = byLabel.get(k)?.unit;
      return unit ? `${k}: ${v} ${unit}` : `${k}: ${v}`;
    })
    .join("  ·  ");
}

export function formatLogLine(vehicle, log) {
  return formatLogValues(log?.values, fieldsForLog(vehicle, log));
}
